hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AuthenticationBaseMessageTask.java

```diff
@@ -314,21 +286,67 @@ public Permission getRequiredPermission() {
         return null;
     }
 
-    private class OperationInfo {
-        private final Member member;
-        private final Future future;
+    private final class MultiTargetCallback {
 
-        public OperationInfo(Future future, Member member) {
-            this.future = future;
-            this.member = member;
+        private final Collection<Member> targets;
+        private final Map<Member, Object> results;
+
+        private MultiTargetCallback(Collection<Member> targets) {
+            this.targets = synchronizedSet(new HashSet<Member>(targets));
+            this.results = new ConcurrentHashMap<Member, Object>(targets.size());
+        }
+
+        public void notify(Member member, Object result) {
+            if (targets.remove(member)) {
+                results.put(member, result);
+            } else {
+                if (results.containsKey(member)) {
+                    logger.warning("Duplicate response from -> " + member);
+                }
+                logger.warning("Unknown target! -> " + member);
+            }
+            if (!targets.isEmpty()) {
```

Comment:
There is a potential race here. If 2 responses are returned at the same time, it could be that both responses trigger the removal of the item on target, but one of the responses has not put the item yet in the 'results'. So a partially filled 'results' could be send to the client.